### ISSUE TITLE
[FEA] Enable structs of lists in the new row lexicographic comparator

### ISSUE THREAD

**Issue by vyasr** [rapidsai/cudf#11672]
**Is your feature request related to a problem? Please describe.**
#11129 enables comparisons of list dtypes with the new comparator. In principle, this code should (in conjunction with prior work like #10164) be sufficient to enable structs of lists (but not lists of structs) via the struct decomposition approach used to preprocess structs. Currently, we forbid comparison of lists of structs in the comparator. During development of #11129 I attempted to remove this restriction, but I observed that this led to [unexpected test failures of [groupby_structs_test/lists_are_unsupported](https://gpuci.gpuopenanalytics.com/job/rapidsai/job/gpuci/job/cudf/job/prb/job/cudf-gpu-test/CUDA=11.0,GPU_LABEL=driver-450,LINUX_VER=centos7,PYTHON=3.8/10269/). These failures were not caused by a failure of `EXPECT_THROW` (which would be natural since I removed a restriction), but were actual failures in the computed outputs. I was unable to reproduce these failures locally, indicating that there is perhaps some unknown indeterminacy in the ordering of some step in the comparison.

**Describe the solution you'd like**
The struct of list restriction in [`check_lex_compatibility`](https://github.com/rapidsai/cudf/blob/branch-22.10/cpp/src/table/row_operators.cu) should be removed and [the failing test](https://github.com/rapidsai/cudf/blob/branch-22.10/cpp/tests/groupby/structs_tests.cpp#L318) should be fixed such that structs of lists are properly supported.

**Comment by ttnghia** [comment]
For comparing lists, we need to generate dremel data. However:
 * Dremel data is generated only for lists columns that are not children of any other column: https://github.com/rapidsai/cudf/blob/9c9dd54f5ec6834ed87f26828c5e3878ced39b2d/cpp/src/table/row_operators.cu#L265-L270
 * On device, dremel data is considered for comparison only in the same situation: https://github.com/rapidsai/cudf/blob/9c9dd54f5ec6834ed87f26828c5e3878ced39b2d/cpp/include/cudf/table/experimental/row_operators.cuh#L559-L567
 * Even that, structs column cannot have lists children. Otherwise, we will encounter compiler stack overflow with device code due to type dispatcher. The current code only processes child structs iteratively while ignoring lists (`dispatch_void_if_nested`): https://github.com/rapidsai/cudf/blob/9c9dd54f5ec6834ed87f26828c5e3878ced39b2d/cpp/include/cudf/table/experimental/row_operators.cuh#L409-L413

So in order to fix this completely, we need both to generate and consider dremel data for nested lists columns. I'm not sure about the complexity of such work, so it should be done in a separate follow up PR.

Update: I've successfully fixed the dremel issue above. However, that may require changing to the function `decompose_structs` which is also used in equality comparators. Such changes may affect comparison performance. Overall, too much risky for a limited time budget. So the issue needs to be addressed separately and I'll post a PR for it after this one merged whenever I can.

**Comment by ttnghia** [comment]
Fixing dremel issue is easy, just a few LOC. For fixing the type dispatcher, I can't find any way to avoid using such `dispatch_void_if_nested` that does not lead to compiler stack overflow. Thus, I turned to a different approach: Keep using `dispatch_void_if_nested` as currently but change `decompose_structs` such that:
 * For structs-of-lists column, instead of always output a structs column with only one child (such as `struct<list>`), the function will output a structs column having 0 child (`struct<>`). This structs column still has the same null mask as the original input.
 * The change above will be applied only for the input structs column with the first child is lists column.

By doing so, we can avoid dispatching `element_comparator` for lists column inside device kernel for structs comparison. I've finished implementation for this which can produce correct results for simple input. I'll test more and will post a complete PR with benchmark later when I have time.
